fastpair/rust/demo/rust/src/api.rs

```diff
@@ -0,0 +1,222 @@
+use std::{collections::HashMap, sync::RwLock, time::Duration};
+
+use bluetooth::{
+    api::{BleAdapter, ClassicDevice},
+    BleDataTypeId, ClassicAddress, PairingResult, Platform,
+};
+use flutter_rust_bridge::StreamSink;
+use futures::executor;
+use tracing::{info, warn};
+use ttl_cache::TtlCache;
+
+use crate::advertisement::{FpPairingAdvertisement, ModelId};
+
+// Sends a device name to Flutter via `StreamSink` FFI layer.
+static DEVICE_STREAM: RwLock<Option<StreamSink<Option<[String; 2]>>>> = RwLock::new(None);
+
+// Saves the currently displayed device's advertisement, to be used for pairing.
+static CURR_DEVICE_ADV: RwLock<Option<FpPairingAdvertisement>> = RwLock::new(None);
+
+// Temporarily restricts which model IDs can be displayed.
+static MODEL_ID_BLACKLIST: RwLock<Option<TtlCache<ModelId, ()>>> = RwLock::new(None);
+
+// How long entries should blacklisted for for.
+const TTL_BLACKLIST: Duration = Duration::from_secs(10);
+
+/// Updates the device name as displayed by Flutter.
+#[inline]
+async fn update_displayed_device(best_adv: Option<FpPairingAdvertisement>) {
+    match best_adv {
+        Some(best_adv) => {
+            match DEVICE_STREAM.read().unwrap().as_ref() {
+                Some(stream) => {
+                    stream.add(Some([
+                        best_adv.name().to_string(),
+                        best_adv.image_url().to_string(),
+                    ]));
+                }
+                None => info!("Name stream is None"),
+            }
+            let mut curr_adv = CURR_DEVICE_ADV.write().unwrap();
+            *curr_adv = Some(best_adv);
+        }
+        None => (),
+    }
+}
+
+/// Sets up necessary constructs to maintain a TTL blacklist of model IDs.
+#[inline]
+fn init_cache() {
+    let mut cache = MODEL_ID_BLACKLIST.write().unwrap();
+    *cache = Some(TtlCache::new(16));
+}
+
+/// Sets up initial constructs and infinitely polls for advertisements.
+pub fn init() {
+    let run = async {
+        info!("start making adapter");
+
+        let mut adapter = Platform::default_adapter().await.unwrap();
+        adapter.start_scan().unwrap();
+
+        init_cache();
+
+        let mut latest_advertisement_map: HashMap<String, FpPairingAdvertisement> = HashMap::new();
+        let mut best_adv: Option<FpPairingAdvertisement> = None;
+        let datatype_selector = vec![BleDataTypeId::ServiceData16BitUuid];
+
+        loop {
+            // Retrieve the next received advertisement.
+            let advertisement = adapter
+                .next_advertisement(Some(&datatype_selector))
+                .await
+                .unwrap();
+
+            // Analyze service data sections.
+            for service_data in advertisement.service_data_16bit_uuid().unwrap() {
```

Comment:
this logic could be encapsulated in a struct that's fed new advertisements and emits the new best one, if any